src/main/java/org/opensearch/security/configuration/DlsFlsValveImpl.java

```diff
@@ -437,7 +437,7 @@ private void setFlsHeaders(EvaluatedDlsFlsConfig dlsFls, ActionRequest request)
             } else {
 
                 if (threadContext.getHeader(ConfigConstants.OPENDISTRO_SECURITY_MASKED_FIELD_HEADER) != null) {
-                    if (!maskedFieldsMap.equals(Base64Helper.deserializeObject(threadContext.getHeader(ConfigConstants.OPENDISTRO_SECURITY_MASKED_FIELD_HEADER)))) {
+                    if (!maskedFieldsMap.equals(Base64Helper.deserializeObject(threadContext.getHeader(ConfigConstants.OPENDISTRO_SECURITY_MASKED_FIELD_HEADER), threadContext.getTransient(ConfigConstants.USE_JDK_SERIALIZATION)))) {
```

Comment:
`ConfigConstants.USE_JDK_SERIALIZATION)` is a setting right? It will be constant once a node is initialized and cannot be changed without a node reboot?

If that's the case is there a way to initialize the `Base64Helper` with the setting and avoid having to pass it in at an argument each time serialize or deserialize is called?

Reply:
I see now how down below if the version is before OpenSearch 3.0.0 its assumed to be JDK serialization, but in >= 3.0.0 it takes in the setting to either continue using JDK serialization or use custom serialization and the logic in `SecurityInterceptor.ensureCorrectHeaders()` is intended to help with backwards compatibility. 

If I am reading it correctly, I can have an OpenSearch 3.0.0 configured with:

```
plugins.security.use_jdk_serialization: true
```

but when it comes to `SecurityInterceptor.ensureCorrectHeaders()` it will always use custom serialization since the node is >= 3.0.0? i.e. this call which does not read in the setting, but only looks at the version is always true if the Node is 3.0.0:

```
final boolean useJDKSerialization = connection.getVersion().before(Version.V_3_0_0);
```

What does this mean for backwards compatibility where there is a mixed cluster during a rolling upgrade?

Reply:
> ConfigConstants.USE_JDK_SERIALIZATION) is a setting right? It will be constant once a node is initialized and cannot be changed without a node reboot?

`ConfigConstants.USE_JDK_SERIALIZATION` is intended to be a transient header and NOT a node setting. Let me break it down for the send and receive path below -

#### Send Request

When sending a request to a remote node, `SecurityInterceptor.sendRequestDecorate()` checks the remote node version and serialises the headers using JDK serialization if the remote node < v3.0.0, otherwise it'll use custom serialization. 

Within `SecurityInterceptor.sendRequestDecorate`, `useJDKSerialization` is used in these two places -
* https://github.com/opensearch-project/security/pull/2802/files#diff-a794e1633c49a387e4883f9dd28f4770a114e3052544ebc1d2673e6b7e1067e4R202
* https://github.com/opensearch-project/security/pull/2802/files#diff-a794e1633c49a387e4883f9dd28f4770a114e3052544ebc1d2673e6b7e1067e4R210

There's no involvement of `ConfigConstants.USE_JDK_SERIALIZATION` here.

#### Receive Request

When receiving a request from remote node within `SecurityRequestHandler.receiveRequestDecorate` I'm able to use the same logic to know how the headers were serialised, but the headers are immutable and we're fetching and deserialising these headers from `ThreadContext` in different classes. I needed a way to make these classes aware of the protocol to be used for deserializing them. To solve this, I use transient header - `ThreadContext.putTransient(ConfigConstants.USE_JDK_SERIALIZATION, useJDKSerialization)`. As per my understanding, transient headers' scope is limited to the request lifecycle on the local node and they're populated within `SecurityRequestHandler.receiveRequestDecorate` for every request that lands on the node.


This all being said, I'm not very sure if `plugins.security.use_jdk_serialization: true` will change the transient headers. 😄